--- src/carbon_integration/osx-integration.h	(before formatting)
+++ src/carbon_integration/osx-integration.h	(after formatting)
@@ -15,28 +15,28 @@
  * Released under GNU GPL, read the file 'COPYING' for more information
  */

+#include "widgets/desktop-widget.h"
 #include <gtkmacintegration/gtkosxapplication.h>
-#include "widgets/desktop-widget.h"

 #ifdef __cplusplus
 extern "C" {
 #endif /* __cplusplus */

 typedef struct _InkscapeOsxIntegration {
-	GtkosxApplication *pGtkOsxApplication;
-	GtkWidget *menubar;
-	GtkWidget *about_item, *prefs_item;
-	SPDesktopWidget *dtw_infocus;
-
+    GtkosxApplication *pGtkOsxApplication;
+    GtkWidget *menubar;
+    GtkWidget *about_item, *prefs_item;
+    SPDesktopWidget *dtw_infocus;
+
 } InkscapeOsxIntegration;


 extern InkscapeOsxIntegration sInkscapeOsxIntegration;
 extern GtkWidget *osx_integration_menuitems[4];

-void OSX_set_menubar( SPDesktopWidget *menubar);
-void   init_Mac_OSX_Integration();
-void   init_Mac_OSX_Integration_complete();
+void OSX_set_menubar(SPDesktopWidget *menubar);
+void init_Mac_OSX_Integration();
+void init_Mac_OSX_Integration_complete();

 #ifdef __cplusplus
 }

--- src/carbon_integration/osx-glue.m	(before formatting)
+++ src/carbon_integration/osx-glue.m	(after formatting)
@@ -1,20 +1,23 @@
+#include <CoreServices/CoreServices.h>
 #include <Foundation/Foundation.h>
-#include <CoreServices/CoreServices.h>

-char *convert_mimetype(const char *mimetype) {
-	char *ret;
-	NSString *str = [[NSString alloc] initWithUTF8String:mimetype];
-        CFStringRef uti = UTTypeCreatePreferredIdentifierForTag(kUTTagClassMIMEType, (CFStringRef)str, NULL);
-        ret =  [(NSString *)uti UTF8String];
-        [str release];
-        return ret;
+char *convert_mimetype(const char *mimetype)
+{
+    char *ret;
+    NSString *str = [[NSString alloc] initWithUTF8String:mimetype];
+    CFStringRef uti = UTTypeCreatePreferredIdentifierForTag(kUTTagClassMIMEType, (CFStringRef)str, NULL);
+    ret = [(NSString *)uti UTF8String];
+    [str release];
+    return ret;
 }


-char *convert_uti(const char *uti) {
-	char *ret = NULL;
-	NSString *str = (NSString *)UTTypeCopyPreferredTagWithClass((CFStringRef)[[NSString alloc] initWithUTF8String:uti], kUTTagClassMIMEType);
-	ret = [str UTF8String];
-	[str release];
-	return ret;
+char *convert_uti(const char *uti)
+{
+    char *ret = NULL;
+    NSString *str = (NSString *)UTTypeCopyPreferredTagWithClass((CFStringRef)[[NSString alloc] initWithUTF8String:uti],
+                                                                kUTTagClassMIMEType);
+    ret = [str UTF8String];
+    [str release];
+    return ret;
 }
 
--- src/carbon_integration/osx-integration.cpp	(before formatting)
+++ src/carbon_integration/osx-integration.cpp	(after formatting)
@@ -13,83 +13,73 @@
  */
 #ifdef __APPLE__

+#include "osx-integration.h"
 #include <gtkmacintegration/gtkosxapplication.h>
-#include "osx-integration.h"

+#include <file.h>
 #include <ui/interface.h>
-#include <file.h>

 #ifdef __cplusplus
 // extern "C" {
 #endif /* __cplusplus */

-static gboolean
-osx_quartz_quit (GtkosxApplication *app, gpointer data);
-static gboolean
-osx_quartz_block_termination (GtkosxApplication *app, gpointer data);
+static gboolean osx_quartz_quit(GtkosxApplication *app, gpointer data);
+static gboolean osx_quartz_block_termination(GtkosxApplication *app, gpointer data);

-InkscapeOsxIntegration sInkscapeOsxIntegration = {0};
+InkscapeOsxIntegration sInkscapeOsxIntegration = { 0 };

 GtkWidget *osx_integration_menuitems[4];

-static gboolean
-osx_quartz_block_termination (GtkosxApplication *app, gpointer data)
+static gboolean osx_quartz_block_termination(GtkosxApplication *app, gpointer data) { return !sp_ui_close_all(); }
+
+
+static gboolean osx_quartz_quit(GtkosxApplication *app, gpointer data)
 {
-	return !sp_ui_close_all();
+    sp_ui_close_all();
+    g_print("Got here");
+    gtk_main_quit();
+    return FALSE;
+}
+
+static gboolean osx_open_file(GtkosxApplication *app, gchar *path, gpointer data)
+{
+    Glib::ustring ustr = Glib::ustring(path);
+    sp_file_open(ustr, NULL, true, false);
+    return true;
 }


-static gboolean
-osx_quartz_quit (GtkosxApplication *app, gpointer data)
+void init_Mac_OSX_Integration()
 {
-	sp_ui_close_all();
-	g_print("Got here");
-	gtk_main_quit();
-	return FALSE;
+    sInkscapeOsxIntegration.pGtkOsxApplication = (GtkosxApplication *)g_object_new(GTKOSX_TYPE_APPLICATION, NULL);
+    // gtkosx_application_set_use_quartz_accelerators(sInkscapeOsxIntegration.pGtkOsxApplication, FALSE);
 }

-static gboolean
-osx_open_file (GtkosxApplication *app, gchar *path ,gpointer data)
+void init_Mac_OSX_Integration_complete()
 {
-	Glib::ustring ustr = Glib::ustring(path);
-	sp_file_open(ustr, NULL, true, false);
-	return true;
+    gtkosx_application_ready(sInkscapeOsxIntegration.pGtkOsxApplication);
+    g_signal_connect(sInkscapeOsxIntegration.pGtkOsxApplication, "NSApplicationBlockTermination",
+                     G_CALLBACK(osx_quartz_block_termination), NULL);
+    g_signal_connect(sInkscapeOsxIntegration.pGtkOsxApplication, "NSApplicationWillTerminate",
+                     G_CALLBACK(osx_quartz_block_termination), NULL);
+    g_signal_connect(sInkscapeOsxIntegration.pGtkOsxApplication, "NSApplicationOpenFile", G_CALLBACK(osx_open_file),
+                     NULL);
+    gtk_accel_map_save("/tmp/inkscape.map");
 }

-
-void   init_Mac_OSX_Integration() {
-	sInkscapeOsxIntegration.pGtkOsxApplication  = (GtkosxApplication*) g_object_new(GTKOSX_TYPE_APPLICATION, NULL);
-	//gtkosx_application_set_use_quartz_accelerators(sInkscapeOsxIntegration.pGtkOsxApplication, FALSE);
-}
+void OSX_set_menubar(SPDesktopWidget *dtw)
+{
+    gtkosx_application_set_menu_bar(sInkscapeOsxIntegration.pGtkOsxApplication, GTK_MENU_SHELL(dtw->menubar));

-void   init_Mac_OSX_Integration_complete() {
-	gtkosx_application_ready(sInkscapeOsxIntegration.pGtkOsxApplication);
-	g_signal_connect(sInkscapeOsxIntegration.pGtkOsxApplication, "NSApplicationBlockTermination",
-	                 G_CALLBACK(osx_quartz_block_termination), NULL);
-	g_signal_connect(sInkscapeOsxIntegration.pGtkOsxApplication, "NSApplicationWillTerminate",
-	                 G_CALLBACK(osx_quartz_block_termination), NULL);
-	g_signal_connect(sInkscapeOsxIntegration.pGtkOsxApplication, "NSApplicationOpenFile",
-	                 G_CALLBACK(osx_open_file), NULL);
-	gtk_accel_map_save("/tmp/inkscape.map");
-}
-
-void OSX_set_menubar(SPDesktopWidget *dtw) {
-        gtkosx_application_set_menu_bar(sInkscapeOsxIntegration.pGtkOsxApplication, GTK_MENU_SHELL (dtw->menubar));
-
-        if (dtw->about_item) {
-            GtkWidget *sep = gtk_separator_menu_item_new();
-            gtkosx_application_insert_app_menu_item  (sInkscapeOsxIntegration.pGtkOsxApplication,
-                                                      dtw->about_item,
-                                                      0);
-            g_object_ref(sep);
-            gtkosx_application_insert_app_menu_item  (sInkscapeOsxIntegration.pGtkOsxApplication, sep, 1);
-        }
-        if (dtw->prefs_item) {
-            gtkosx_application_insert_app_menu_item  (sInkscapeOsxIntegration.pGtkOsxApplication,
-                                                      dtw->prefs_item,
-                                                      2);
-        }
-
+    if (dtw->about_item) {
+        GtkWidget *sep = gtk_separator_menu_item_new();
+        gtkosx_application_insert_app_menu_item(sInkscapeOsxIntegration.pGtkOsxApplication, dtw->about_item, 0);
+        g_object_ref(sep);
+        gtkosx_application_insert_app_menu_item(sInkscapeOsxIntegration.pGtkOsxApplication, sep, 1);
+    }
+    if (dtw->prefs_item) {
+        gtkosx_application_insert_app_menu_item(sInkscapeOsxIntegration.pGtkOsxApplication, dtw->prefs_item, 2);
+    }
 }

 #ifdef __cplusplus
